File: python/sliding_window/leet03.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_length = 0
        left = 0

        seen_chars = {}

        for right in range(0, len(s)):
            char = s[right]
            if char in seen_chars:
                left = max(left, seen_chars[char]+1)

            seen_chars[char] = right
            max_length = max(max_length, right-left+1)

        return max_length


if __name__ == '__main__':
    s = Solution()
    val = "pwwkew"
    s.lengthOfLongestSubstring(val)

File: java/leetcode-practice/src/test/java/com/mamta/leetcode/slidingwindow/leet03.java
package com.mamta.leetcode.slidingwindow;

import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

public class leet03 {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
       int left = 0;
       int maxLength = 0;

       for(int right=0; right<s.length(); right++) {

           Character c = s.charAt(right);

           if(map.containsKey(c)) {
               left = Math.max(left, map.get(c)+1);
           }

           map.put(c, right);
           maxLength = Math.max(maxLength, right-left+1);
       }
       return maxLength;
    }

    @Test
    public void solution() {
        String s = "pwwkew";
        System.out.println(lengthOfLongestSubstring(s));
    }
}
